Edit position and more

wbazant opened the pull request:
This is to allow editing position of the location, and make some (many) associated changes.

Sorry for no individual commits - I dropped the commit history for editing the location position because there were many of them, and they didn't apply cleanly to the updated branch.

What's here:
- a visual element in the location form, clickable on mobile
- changes to the location pins (see https://github.com/falling-fruit/falling-fruit-web/pull/404#issuecomment-2205772656)
- on desktop, a pin appears during "edit location", that can be dragged around closes #409 
- on mobile, there's a new 'edit location position" with navs back and forth
- mechanics of saving the form during back and forth - Closes #405
- some more refactoring - closes #406 
- zoom to location on edit, improve zoom in / out, remove zooming back, disable add location button - partially addresses #398
- fix tiny bug closes #417

ezwelty commented:
@wbazant Thanks for this. I found some faster internet and was able to take it for a spin. Comments and bug reports:

* [desktop] No marker is visible during add location. Did it get lost in the move to a movable edit-location marker?
* [desktop] Once the marker is moved, the new position is not cleared on back or cancel, so that when edit location is reinitiated, it begins with the new position. I would expect it to restart at the current position?
* [mobile] Shouldn't the position link be orange, since that is the color used for most clickable elements?
* [mobile] This sequence: Edit location > Edit position > Edit location > Back/Cancel... results in a weird half-loaded view location with the drawer half-open but no map loaded. I notice that the back/cancel URL (e.g. `/locations/1990474`) loses the map center and zoom after a round trip through Edit position. More generally, a URL of this form does not load properly.
* I guess #417 is not fixed on mobile, in the sense that clicking back does not go to a map centered on the associated location. On both mobile and desktop, ideally clicking back would go to the associated location rather than the bare map, but maybe this is a distraction.

wbazant commented:
Thanks! Fixed those:
[desktop] No marker is visible during add location. Did it get lost in the move to a movable edit-location marker?
[desktop] Once the marker is moved, the new position is not cleared on back or cancel, so that when edit location is reinitiated, it begins with the new position. I would expect it to restart at the current position?
[mobile] Shouldn't the position link be orange, since that is the color used for most clickable elements?
![Screenshot from 2024-07-07 11-00-56](https://github.com/falling-fruit/falling-fruit-web/assets/16976512/2f824f25-0d44-47fd-ad97-88f911cefdd4)
Thanks for the suggestion to make it orange, I tried to earlier make it a proper button and it didn't look very good, but I think orange fits with the rest of the form better than underline.

I started investigating that bug with location/:locationId on mobile, it's to do with my change to start from null view and not the initial one - the map loads at zero pixels size when it happens.

wbazant commented:
I fixed it by adding an extra navigation after /locations/:locationId is loaded - I couldn't coax the map into repairing itself after it loaded itself to zero pixels. Also I made the problem not show up by keeping the map state in the URL.

Made an issue #420 for going back to location after going back from the review page

wbazant commented:
Now also closes #420

wbazant commented:
Now also closes #397 
![Screenshot from 2024-07-08 21-58-02](https://github.com/falling-fruit/falling-fruit-web/assets/16976512/145d2ab8-5d5c-4ad0-90a5-c725b84cc297)

wbazant commented:
#398 needs more work - on mobile we have to zoom in and center on location because of the UX involving a central unmovable pin (so that dragging the map changes proposed location) but since I disabled the Add location button, on desktop we shouldn't manage the map for the users at all.

Also, centering shouldn't happen when loading /edit/position, we want URLs like
http://localhost:3000/locations/1830121/edit/position/@55.83062349376405,-4.273989356893054,16z
to load correctly (showing a moved position) so that the round trip to settings doesn't reset the pin.

This doesn't ruin the pull request, there's still other bugs in it to be found and reported :) but I'm running out of beans for today so writing this down for reference!

wbazant commented:
Now also closes #374

wbazant commented:
I can still do some more work on top of this PR - state of the location drawer and zoom buttons makes sense - but I'm also quite happy with what I changed here so far, and I squashed the bugs I could find, so I'm unmarking it as draft.

wbazant commented:
Now also adds zoom buttons - I think we don't have a ticket but proposed during #398 - and helps with #214.

wbazant commented:
Now also closes #423

wbazant commented:
Now also closes #418 and #425

wbazant commented:
I've started a series of changes to use the google map API directly instead of state.map.view. It's in general producing improvements - we get nice animations, it's faster, we can rely on map.getBounds() returning bounds, etc. 

The end result could be to remove state.map.view and state.list.view, or I'll just fix related bugs until there are no more, but it's a big change and I've only just completed parts of it. I don't want to sit on the changes so I'll push it here, mark the PR as draft, and I'll get it completed over time.

wbazant commented:
I partially addressed #426 - except I saw the place go away again so still flakey.

Closes #387 - not using the view.bounds but googleMap.getBounds() so I expect the problem to stop appearing.

Partial work for #214 - the ticket also mentions "Loading the type filter is slow, and impacts the performance of the map." and I'm not achieving this, but "Panning and zooming the Google Map component is often sluggish" is good now.

Relates to #389 but I've seen it not work correctly, I'll need to revisit the interaction of map, list, and filter as part of this change.


Also, I expect high severity bugs like e.g. tab sometimes crashing on load that I've just seen.

wbazant commented:
I'm gonna bravely make this ready for review after a streak of successes: I've fixed the crash, and then was able to remove all remaining uses of map.view.

wbazant commented:
TODO correct these usages:
```
wbazant@wbazant-19-01:~/dev/falling-fruit-web$ ag view.zoom
docs/map-view-redux.md
31:  zoom={view.zoom}

src/components/entry/EntryOverview.js
75:    if (view.zoom < MIN_LOCATION_ZOOM) {

src/components/connect/ConnectNewLocation.js
25:          zoom: Math.max(view.zoom, isDesktop ? 0 : 16),

src/utils/appUrl.js
59:  )},${view.zoom}z`

src/redux/viewChange.js
39:  const screenDist = dist * Math.pow(Math.pow(2, newView.zoom), 2)
45:    newView.zoom,
```

wbazant commented:
I've made street view into its own page - /locations/:locationId/panorama - which allows linking to it and solves some problems. Closes #427 and #428.

I've just noticed a bug in the PR - when I resize the window, center shifts to the default location - to do with using view from google maps, and the callback changing the bounds maybe not having a center. I'll fix it soon

wbazant commented:
#426 is now fixed better - I was looking at the previous incomplete solution, removed a call to GoogleMapLoader, and it somehow helped usePlacesAutocomplete to work better. We still sync the value to redux and back but don't need the suggestions that I tried putting in there earlier.

wbazant commented:
I was checking if #389 improved, and it's not yet done. I've found these two issues:
- on refreshing a list URL like http://localhost:3000/list/@55.8268570,-4.2611208,14z we get an empty list, and the search is not clickable
- when searching in the list view, we get very low zoom values, for example "G429DN" gives a zoom of 8 (14 on map)

wbazant commented:
This PR now also closes #347.

Spotted a bug: on mobile, when in /list and going to account, the locations aren't right (but going back and forth between map and list works). Probably to do with my changes to the list slice.

wbazant commented:
This PR now also closes #379

wbazant commented:
This PR now also closes #314
![Screenshot from 2024-07-29 11-28-12](https://github.com/user-attachments/assets/92e4c1b3-d4ea-463c-83ea-cdc81726c13d)

wbazant commented:
This PR now also closes #353

wbazant commented:
Panorama still sometimes doesn't load: I've accidentally found http://localhost:3000/locations/1826768/panorama shows google maps and no error but the original http://localhost:3000/locations/1756237/panorama/ does the same

wbazant commented:
This PR now also closes #434

ezwelty commented:
Wow, @wbazant you've been busy! It preparation for out call tomorrow, I went through every mentioned issue and feature and wrote out all of my observations below.

## Map zoom and pan

- The (-) zoom button is disabled from z <= 4 even though the map can be zoomed out to z = 3 by scrolling. Note that I don't see any reason to set a minimum zoom level other than 0. Why not let users fit the whole world on their screen?
- The (+) zoom button is only disabled (at z = 22) for the map layer. It remains active (but has no effect) for the terrain and satellite layers.
- Map pan and zoom seems to have gotten even more laggy. It is noticeable even with an empty map, but worse when zoomed way out, and far worse on desktop than on mobile (or on desktop in add/edit location mode), which suggests that map click events are blocked by API calls and rendering of the type tree.

## Map markers

- [mobile] The "select location" marker is not removed from the map (and the "add location" marker remains hidden) when the drawer is closed by swiping it closed or clicking elsewhere on the map.
- [mobile] Q: Maybe silly, but it feels unclear to me, when canceling "edit location position", whether the change I made was in fact canceled. Would it help to make the map recenter on the original position briefly before returning to the "edit location" form as visual confirmation of the cancelation?
- [desktop] The infobox for the "add location" position marker reopens automatically on each map change (I would expect it to remain closed if I clicked the 'x' to close it). Clicking the marker does not reopen the infobox (I would expect it to). The infobox for the "edit location" marker does not reopen at all.
- [desktop] Moving the "add location" and "edit location" markers around tends to temporarily highlight the text of all the surrounding location labels.
- Q: Should we use the "edit location" marker styling for "add location"? i.e. the same orange marker? The blue marker often has poor contrast with the basemap.

## Page navigation

- [mobile] Going back from "edit location" (even if no changes were made and the "edit position" mode was not initiated) results in the map centered on the location and no other locations loaded. Loading other locations is triggered only once the map is changed.
- [mobile] #347: This seems to be broken in a new way. http://localhost:3000/list/@40.0405403,-105.2632396,21z loads the list, but switching to the map view does not load markers (weirdly I only see one basemap tile loaded, the rest load only after the map is moved), and moving the map does not update the map center and zoom in the URL, such that clicking back to the list always shows the original list.

## Panorama (Street) view

- [mobile] The [x] button is in an odd position (lower than expected).
- [mobile] Closing the panorama goes back to the map even if it was opened from the list view.
- Q: Subsequent loads of the panorama sometimes shows an animation to the new location, even though this is irrelevant for the user. Can we force a fade-in instead?
- Q: Should the link to the panorama be disabled / greyed out once we've determined that there is no panorama available (and toasted accordingly)?

## Search bar

- #426: I find the crosshair icon for the selected search result confusing because we already use that symbol for device location. How about either making all of the others a simple marker icon (without the filled circle, see maps.google.com) and the selected one with a circle to indicate selection? Or use a clock icon (i.e. "previously selected"), which then better works with the following suggestion:
- Q: Probably a crosshair/label should not be added if the geocoded location type is not a point – i.e. `APPROXIMATE` or `GEOMETRIC_CENTER` (see https://developers.google.com/maps/documentation/geocoding/requests-geocoding)? It feels funny to search for a country or city and end up with a crosshair on the map. Or alternatively – the strategy used by the current website – only show the crosshair when resulting map zoom > 12.
- Q: Clicking on a search result, NOT clicking anywhere, then clicking in the search bar again does not show the search results (I guess because it never loses focus), but somehow it feels like it should?

wbazant commented:
Thanks, that's a lot of good feedback! I think I've also just managed to finish a feature I started locally - modal form to add a new type - I'll hold on to it and we can discuss what to do with it all.

wbazant commented:
Now also closes #305 
![Screenshot from 2024-08-13 17-27-29](https://github.com/user-attachments/assets/f8c189ec-6c8a-4427-8423-dcbf232d6bbb)

I'll address your comments/questions next @ezwelty before I start any more new features! I just had the code for this one almost finished.

wbazant commented:
> Map zoom and pan
>    The (-) zoom button is disabled from z <= 4 even though the map can be zoomed out to z = 3 by scrolling. Note that I don't see any reason to set a minimum zoom level other than 0. Why not let users fit the whole world on their screen?
>    The (+) zoom button is only disabled (at z = 22) for the map layer. It remains active (but has no effect) for the terrain and satellite layers.
>    I'll note that it would straightforward to add Google Maps zoom controls back to the map, if this is easier than rolling up our own. See https://developers.google.com/maps/documentation/javascript/controls#maps_control_disableUI-javascript
 >   Map pan and zoom seems to have gotten even more laggy. It is noticeable even with an empty map, but worse when zoomed way out, and far worse on desktop than on mobile (or on desktop in add/edit location mode), which suggests that map click events are blocked by API calls and rendering of the type tree.

I had a play with the default zoom, it doesn't match the UI very well:
![Screenshot from 2024-08-14 15-11-02](https://github.com/user-attachments/assets/f23c9918-37d8-445a-bec7-e6e55fab78f0)
and I see how to position it, but not how to style it. I added the shading it has to our buttons.

I'm not sure why the minimum is 4.  but the app doesn't let me view lower zooms by scrolling, or going to them with a URL. How do you get to a three?

Setting the maximum correctly would require an extra API call and it might complicate the current performance issue so I'd leave it for later, issue #436

wbazant commented:
>[mobile] The "select location" marker is not removed from the map (and the "add location" marker remains hidden) when the drawer is closed by swiping it closed or clicking elsewhere on the map.
Fixed by improving the logic for connecting URL changes to redux states (the src/components/connect stuff), thanks for pointing it out!

>[mobile] Q: Maybe silly, but it feels unclear to me, when canceling "edit location position", whether the change I made was in fact canceled. Would it help to make the map recenter on the original position briefly before returning to the "edit location" form as visual confirmation of the cancelation?
Thanks, I remember going back and forth when QAing to check the cancel really cancelled. I've added a bit of logic to wait before cancelling, but only if the user displaced the marker.


> [desktop] The infobox for the "add location" position marker reopens automatically on each map change (I would expect it to remain closed if I clicked the 'x' to close it). Clicking the marker does not reopen the infobox (I would expect it to). The infobox for the "edit location" marker does not reopen at all.
Fixed first one - map is no longer reopening - and added the behaviour about clicking, thanks! It will probably help establish that we want the user to drag, not click. 


>[desktop] Moving the "add location" and "edit location" markers around tends to temporarily highlight the text of all the surrounding location labels.
Made label text not highlightable

> Q: Should we use the "edit location" marker styling for "add location"? i.e. the same orange marker? The blue marker often has poor contrast with the basemap.
Hm, I like that they're different elements - I think the two UXes should be distinguishable even though we use the same elements, and we already use blue labels on the map. Not hugely convinced by the suggestion but not opposed to it either. Also, I've spotted something else - Editing Location has a header over the form, but adding location doesn't - and added the extra header.

wbazant commented:
>[mobile] Going back from "edit location" (even if no changes were made and the "edit position" mode was not initiated) results in the map centered on the location and no other locations loaded. Loading other locations is triggered only once the map is changed.

Fixed, thanks!

ezwelty commented:
> I'm not sure why the minimum is 4. but the app doesn't let me view lower zooms by scrolling, or going to them with a URL. How do you get to a three?

I just zoom out by scrolling with my touchpad? Does this not work for you? http://beta.fallingfruit.org/map/@36.8119176,-76.4573812,3z

Anyway, I just lowered the max zoom to 1 in https://github.com/falling-fruit/falling-fruit-web/pull/419/commits/3f8957fee3d45a8523872f1101cfd95e55cd5d2e (0 should work in theory, but google-map-react doesn't seem to like it, and 1 is plenty).

ezwelty commented:
> I added the shading it has to our buttons.

You mean the shading on hover? I notice that on mobile, the shading just sticks after the button is clicked. Until the other button is clicked or one navigates away from and back to the map. Is that a bug or intentional?

wbazant commented:
I forgot to update the thread when I made the code changes but this PR also closes #427 and closes #428, partially addresses #425 and partially addresses #426. Thanks for pointing out @ezwelty!

ezwelty commented:
@wbazant I noticed a new issue introduced by this PR: Unnecessary API calls on first load.

* Calls `/types` twice in quick succession (rather than just once)
* [mobile] Calls `/type/counts` on first load (rather than not at all), even though it is not needed on mobile (`/type/counts` is only called if the type filter window is open).

wbazant commented:
> You mean the shading on hover? I notice that on mobile, the shading just sticks after the button is clicked. Until the other button is clicked or one navigates away from and back to the map. Is that a bug or intentional?

bug! I've changed the code to only have the hover effect on desktop.

wbazant commented:
>    [mobile] The [x] button is in an odd position (lower than expected).

I agree! We're tracking this improvement as part of the #416 issue 

>    [mobile] Closing the panorama goes back to the map even if it was opened from the list view.

We've got an issue for this bug, #360


>    Q: Subsequent loads of the panorama sometimes shows an animation to the new location, even though this is irrelevant for the user. Can we force a fade-in instead?

I saw that too, I think I've not caused it since I've seen it before doing the changes, and then convinced myself it might not be there any more with the new implementation. I agree the animation is not desirable - I've made an issue for it, #437


    Q: Should the link to the panorama be disabled / greyed out once we've determined that there is no panorama available (and toasted accordingly)?

Thanks, implemented! info on panorama being missing for a location could even come from the back-end or we could crowdsource it from users (website posts to the backend etc. ) and we could have scheduled jobs that check whether the panorama is still missing but it doesn't seem worthwhile, while greying out the link was easy and it looks like the app knows what it's doing.

wbazant commented:
>https://github.com/falling-fruit/falling-fruit-web/issues/426: I find the crosshair icon for the selected search result confusing because we already use that symbol for device location. How about either making all of the others a simple marker icon (without the filled circle, see maps.google.com) and the selected one with a circle to indicate selection? Or use a clock icon (i.e. "previously selected"), which then better works with the following suggestion:
Thanks, I agree with the suggestion! I removed the icon so we're just highlighting the color.

> Q: Probably a crosshair/label should not be added if the geocoded location type is not a point – i.e. APPROXIMATE or GEOMETRIC_CENTER (see https://developers.google.com/maps/documentation/geocoding/requests-geocoding)? It feels funny to search for a country or city and end up with a crosshair on the map. Or alternatively – the strategy used by the current website – only show the crosshair when resulting map zoom > 12.
Thanks, I agree the crosshair that said e.g. Scotland felt a bit silly. Added a clause to only show crosshair when zoom is over 12

> Q: Clicking on a search result, NOT clicking anywhere, then clicking in the search bar again does not show the search results (I guess because it never loses focus), but somehow it feels like it should?
I think I agree although not feeling strongly about it. I had a go at seeing if I can implement this, and there's an onClick that can be passed to ComboboxInput, but no obvious way to influence the combobox's state.

ezwelty commented:
@wbazant Unfortunately https://github.com/falling-fruit/falling-fruit-web/pull/419/commits/92240994b43e57b00c5628f171d85b91be13fa8e introduces many issues:

* Tree checkboxes do not update style when clicked
* Tree does not render on back from view-location

ezwelty commented:
@wbazant A regression I notice compared to main is that location description form field (and perhaps others) are very laggy.

wbazant commented:
Thanks! I guess it's not even a partial improvement then. I also had a go at upgrading rc-tree-select again, but it doesn't work out of the box, we're using it in a funny way (the component has a search bar for selecting, with a popover - we only render what's meant to be in the popover).

>@wbazant A regression I notice compared to main is that location description form field (and perhaps others) are very laggy.
Yes! it's the worst in the textarea for adding a new type. I was comparing to beta.fallingfruit.org and convinced myself that the production build of react is going to be faster, but that was definitely wishful thinking.

wbazant commented:
@ezwelty check this one out, I replaced the library with custom code and it pretty much works! It's still a bit wonky:
- items don't align in a grid
- you can't scroll through in the desktop layout
- Malus (Crabapple) renders ahead of Malus (Apple) while I'd want ties to be resolved by common name
- Select all doesn't work

but it looks sort of right, actions of checkboxes seem to apply to the map correctly, and the map lag is gone now.

wbazant commented:
This PR now also closes  #441 and partially addresses #440

ezwelty commented:
> check this one out, I replaced the library with custom code and it pretty much works

Wow, the performance improvement is stunning.

A couple more quirks to add to the list:

* Hard to reproduce, but if I click around in the checkboxes, I eventually end up in a broken state where the API calls to `/locations` and `/clusters` is broken by a weird entry in the `types` param: `...,extra-group-value-115,115,...`
* Funny things happen when items get word-wrapped: the checkbox gets squished (rather than staying square) and is vertical-aligned to the center (rather than the top)

![image](https://github.com/user-attachments/assets/48a7fa26-0fd8-4d26-bc60-69666f72c65a)

wbazant commented:
This PR now also closes #373 and makes the tree select much better! I gave parents that don't match the search but only there for context a "disabled" semantic: they're there but can't be used while the search is active.

still some problems with the code:
- I just noticed lost the logic for displaying cultivars with simpler labels of just with cultivar name (I'll fix it)
- the container around the tree used to nicely fill up the space that was remaining for it, in the layout, and now it doesn't want to do that any more. I set it to 55vh temporarily but @ezwelty would you be up for trying to give the TreeSelectView a "100% of remaining space"?

wbazant commented:
This PR now also closes #441

wbazant commented:
This PR now also closes #305 (updated the form and removed the performance issue)

ezwelty commented:
@wbazant Thanks for all your hard work on this (and sorry I didn't swoop in for a CSS fix). I haven't been very present as all my free time recently was used up preparing for the trip that I leave on tomorrow (back mid-September). Will test and contribute more thoroughly when I'm back.

wbazant commented:
@ezwelty sure, thanks for staying in touch! I'll do some more work here on the issues that are light on UX or planning, and we can pick up the rest together when you're back or whenever you have time.

wbazant commented:
This PR now closes #444 , closes #439, closes #415.

Also, I think the earlier work closes #438 , and I'd bravely say it even closes #214 the performance epic after making google maps act faster, and replacing the tree select with a custom one.

wbazant commented:
This PR now partially addresses  #364 ( I still want to fix the lightbox on mobile) , closes #359, closes #360, partially addresses #431 ( there's no more need for fromPage but I still want to remove occurrences from the code base )

wbazant commented:
On mobile, there are some redux issues that I'd like to not forget:
- `types` sometimes is not loaded before getting locations, resulting in a map with no types
- fetching locations list view stopped working, because on switch to list, the googleMap component cleverly resizes itself to zero pixels. It might be actually a google maps update getting better at knowing when it's on the screen, or maybe I've changed the layout slightly. The solution to it should be to put the viewport bounds in redux.

wbazant commented:
#364 lightbox works but you can lock the drawer in top position without triggering an action somehow! with a quick pull all the way up

wbazant commented:
There's now a lightbox on mobile but there's an outstanding bug where you can put the drawer in a top position by moving fast, and not trigger a change in redux.

wbazant commented:
This PR now closes #364

wbazant commented:
We can now load list locations.

Outstanding bug: starting from list page and setting the filter correctly filters the location, but then going to the map results in a map with zero locations drawn on it. There's a special case about filtering inside mapSlice that should probably be revisited

wbazant commented:
This PR now also closes #448

wbazant commented:
This PR now also closes #445

wbazant commented:
This PR now also closes #431

ezwelty commented:
@wbazant Thanks for all this great work. I'm going to merge and make live. A more careful review will follow as new or updated issues.
